rust/src/normalization.rs

```diff
@@ -120,6 +121,294 @@ pub(crate) fn rewrite_rpmdb_timestamps<F: Read + Write + Seek>(rpmdb: &mut F) ->
+
+    #[derive(BinRead, Debug)]
+    #[br(little)]
+    struct BTreeItem {
+        len: u16,
+        item_type: BTreeItemType,
+    }
+
+    #[derive(BinRead, Debug)]
+    #[br(repr=u8)]
+    #[repr(u8)]
+    enum HashItemType {
+        KeyData = 1,
+        Duplicate = 2,
+        Offpage = 3,
+        OffDup = 4,
+    }
+
+    lazy_static! {
+        static ref PROC_SELF_CWD: PathBuf = PathBuf::from("/proc/self/cwd");
+        static ref PROC_SELF_FD: PathBuf = PathBuf::from("/proc/self/fd");
+    }
+
+    pub(super) fn normalize_rpmdb_bdb<P: AsRef<Path>>(
+        rootfs: &openat::Dir,
+        rpmdb_path: P,
+        timestamp: u32,
+    ) -> Result<()> {
+        let rpmdb_path = rpmdb_path.as_ref();
+
+        for entry in rootfs.list_dir(rpmdb_path)? {
+            let entry = entry?;
+
+            match entry.simple_type() {
+                Some(SimpleType::File) => (),
+                _ => continue,
+            };
+
+            if let Some(filename) = entry.file_name().to_str() {
+                if filename.starts_with('.') || filename.starts_with("__db") {
+                    continue;
+                }
+            } else {
+                continue;
+            }
+
+            let path = rpmdb_path.join(entry.file_name());
+
+            let old_digest = sha2::Sha256::digest(
+                Exec::cmd("db_dump")
+                    .args(&[PROC_SELF_CWD.join(&path)])
+                    .cwd(PROC_SELF_FD.join(rootfs.as_raw_fd().to_string()))
+                    .stdout(Redirection::Pipe)
+                    .capture()
+                    .context("pre-normalization data dump")?
+                    .stdout
+                    .as_slice(),
+            );
+
+            {
+                let mut file_id = sha2::Sha256::new();
+                file_id.update(timestamp.to_be_bytes());
+                file_id.update(format!("rpm/{}", entry.file_name().to_str().unwrap()).as_bytes());
+                let file_id = &file_id.finalize()[..20];
+
+                let mut db = rootfs.update_file(&path, 0o644)?;
+                let meta_header: MetaHeader = db.read_le()?;
+
+                match (meta_header.magic, meta_header.page_type) {
+                    (BTREE_MAGIC, PageType::BTreeMeta) => (),
+                    (HASH_MAGIC, PageType::HashMeta) => (),
+                    _ => continue,
+                };
+
+                db.seek(SeekFrom::Start(PAGE_HEADER_MAGIC_OFFSET))?;
+                db.write_all(file_id)?;
+
+                for pageno in 1..meta_header.last_pgno + 1 {
+                    db.seek(SeekFrom::Start((pageno * meta_header.pagesize) as u64))?;
+
+                    let header: PageHeader = db.read_le()?;
+
+                    if header.page_type == PageType::Overflow {
+                        db.seek(SeekFrom::Current(header.hf_offset as i64))?;
+                        let fill_length =
+                            meta_header.pagesize - (PAGE_HEADER_SIZE + header.hf_offset) as u32;
```

Comment:
Let's use e.g. `checked_sub()` for things like this?  i.e. `meta_header.pagesize.checked_sub((PAGE_HEADER_SIZE + header.hf_offset) as u32)`